File: python/zipper/zipper.py
class Zipper:
    def __init__(self, value, parent):
        self._value = value
        self._parent = parent
        self._root = None
        self._left = None
        self._right = None

    @staticmethod
    def from_tree(tree, parent=None, root=None):
        if tree is None:
            return None

        node = Zipper(tree["value"], parent)
        if root is None:
            root = node
        node._root = root
        node._left = Zipper.from_tree(tree["left"], node, root)
        node._right = Zipper.from_tree(tree["right"], node, root)

        return node

    def value(self):
        return self._value

    def set_value(self, value):
        self._value = value
        return self

    def left(self):
        return self._left

    def set_left(self, tree):
        self._left = Zipper.from_tree(tree, self, self._root)
        return self

    def right(self):
        return self._right

    def set_right(self, tree):
        self._right = Zipper.from_tree(tree, self, self._root)
        return self

    def up(self):
        return self._parent

    def to_tree(self):
        return self._root.to_subtree()

    def to_subtree(self):
        return {
            "value": self._value,
            "left": self._left.to_subtree() if self._left else None,
            "right": self._right.to_subtree() if self._right else None
        }

File: haskell/zipper/src/Zipper.hs
module Zipper
  ( BinTree(BT)
  , fromTree
  , left
  , right
  , setLeft
  , setRight
  , setValue
  , toTree
  , up
  , value
  ) where

data BinTree a = BT
  { btValue :: a
  , btLeft  :: Maybe (BinTree a)
  , btRight :: Maybe (BinTree a)
  }
  deriving (Eq, Show)

data Parent a = L a | R a | N
  deriving (Eq, Show)

data Zipper a = Z
  { zParent :: Parent (Zipper a)
  , zTree   :: BinTree a
  }
  deriving (Eq, Show)

fromTree :: BinTree a -> Zipper a
fromTree = Z N

toTree :: Zipper a -> BinTree a
toTree (Z N     tree) = tree
toTree (Z (L z) _   ) = toTree z
toTree (Z (R z) _   ) = toTree z

value :: Zipper a -> a
value = btValue . zTree

left :: Zipper a -> Maybe (Zipper a)
left (  Z _ (BT _ Nothing  _)) = Nothing
left z@(Z _ (BT _ (Just l) _)) = Just (Z (L z) l)

right :: Zipper a -> Maybe (Zipper a)
right (  Z _ (BT _ _ Nothing )) = Nothing
right z@(Z _ (BT _ _ (Just r))) = Just (Z (R z) r)

up :: Zipper a -> Maybe (Zipper a)
up (Z N     _) = Nothing
up (Z (L z) _) = Just z
up (Z (R z) _) = Just z

setValue :: a -> Zipper a -> Zipper a
setValue x (Z N     tree) = Z N (setValue' x tree)
setValue x (Z (L z) tree) = Z (L (setLeft (Just tree') z)) tree'
  where tree' = setValue' x tree
setValue x (Z (R z) tree) = Z (R (setRight (Just tree') z)) tree'
  where tree' = setValue' x tree

setLeft :: Maybe (BinTree a) -> Zipper a -> Zipper a
setLeft l (Z N     tree) = Z N (setLeft' l tree)
setLeft l (Z (L z) tree) = Z (L (setLeft (Just tree') z)) tree'
  where tree' = setLeft' l tree
setLeft l (Z (R z) tree) = Z (R (setRight (Just tree') z)) tree'
  where tree' = setLeft' l tree

setRight :: Maybe (BinTree a) -> Zipper a -> Zipper a
setRight r (Z N     tree) = Z N (setRight' r tree)
setRight r (Z (L z) tree) = Z (L (setLeft (Just tree') z)) tree'
  where tree' = setRight' r tree
setRight r (Z (R z) tree) = Z (R (setRight (Just tree') z)) tree'
  where tree' = setRight' r tree

setValue' :: a -> BinTree a -> BinTree a
setValue' x (BT _ l r) = BT x l r

setLeft' :: Maybe (BinTree a) -> BinTree a -> BinTree a
setLeft' l (BT x _ r) = BT x l r

setRight' :: Maybe (BinTree a) -> BinTree a -> BinTree a
setRight' r (BT x l _) = BT x l r
